guide/src/type-stub.md

```diff
@@ -0,0 +1,77 @@
+# Type stub generation (`*.pyi` files) and introspection
+
+*This feature is still in active development. See [the related issue](https://github.com/PyO3/pyo3/issues/5137).*
+
+PyO3 has a work in progress support to generate [type stub files](https://typing.python.org/en/latest/spec/distributing.html#stub-files).
+
+It works using:
+1. PyO3 macros (`#[pyclass]`) that generate constant JSON strings that are then included in the built binaries by rustc if the `experimental-inspect` feature is enabled.
+2. The `pyo3-introspection` crate that can parse the generated binaries, extract the JSON strings and build stub files from it.
+3. [Not done yet] Build tools like `maturin` exposing `pyo3-introspection` features in their CLI API.
+
+For example the following Rust code
+```rust
+#[pymodule]
+pub mod example {
+    use pyo3::prelude::*;
+
+    #[pymodule_export]
+    pub const CONSTANT: &str = "FOO";
+
+    #[pyclass(eq)]
+    #[derive(Eq)]
+    struct Class {
+        value: usize
+    }
+
+    #[pymethods]
+    impl Class {
+        #[new]
+        fn new(value: usize) -> Self {
+            Self { value }
+        }
+        
+        #[getter]
+        fn value(&self) -> usize {
+            self.value
+        }
+    }
+    
+    #[pyfunction]
+    #[pyo3(signature = (arg: "list[int]") -> "list[int]")]
+    fn list_of_int_identity(arg: Bound<'_, PyAny>) -> Bound<'_, PyAny> {
+        arg
+    }
+}
+```
+will generate the following stub file:
+```python
+import typing
+
+CONSTANT: typing.Final = "FOO"
+
+class Class:
+    def __init__(self, value: int) -> None: ...
+
+    @property
+    def value(self) -> int: ...
+
+    def __eq__(self, other: Class) -> bool: ...
+    def __ne__(self, other: Class) -> bool: ...
+
+def list_of_int_identity(arg: list[int]) -> list[int]: ...
+```
+
+The only piece of added syntax is that the `#[pyo3(signature = ...)]` attribute
+can now contain type annotations like `#[pyo3(signature = (arg: "list[int]") -> "list[int]")]`
+(note the `""` around type annotations).
+This is useful when PyO3 is not able to derive proper type annotations by itself.
```

Comment:
Rather than listing this here, perhaps instead edit the `signature` documentation to say that if the feature is enabled, it accepts type hints (and link to this page)?

Reply:
Done! I have also added links between the new page and the one about type stubs.